crates/sui-types/src/intent.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::batch::AuthorityBatch;
+use crate::messages::{TransactionData, TransactionEffects};
+use crate::messages_checkpoint::CheckpointSummary;
+
+use serde::{Deserialize, Serialize};
+
+/// Current Sui version (for compatibility purposes).
+/// TODO: Increment value when a non-backwards compatible Sui version is launched (i.e. when
+/// changing the format, signature or field ordering of signable types).
+const SUI_COMPATIBILITY_VERSION: u8 = 0;
+const SUI_CHAIN_ID: u8 = 0;
+
+/// This is the only type we should sign per our serialization-handbook to provide domain separation
+/// and avoid accidental serialized-bytes collisions between structs and intents, different versions
+/// of the same struct and chains (mainnet, testnet etc).
+/// Similarly, `SecureIntent` should be preferred when hashing requires domain separation
+/// guarantees.
+pub trait SecureIntent: Serialize + serde::de::DeserializeOwned + private::SealedIntent {}
+
+/// Struct required when Sui repository-version and chainID are required for domain
+/// separation (i.e. when something is sign-able).
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct VersionPlusChainId {
+    version: u8,
+    chain_id: u8,
+}
+
+impl Default for VersionPlusChainId {
+    // TODO: read these values from some config, genesis or epoch.
+    fn default() -> Self {
+        VersionPlusChainId {
+            version: SUI_COMPATIBILITY_VERSION,
+            chain_id: SUI_CHAIN_ID,
+        }
+    }
+}
+
+/// `IntentScope` is required to guarantee two different intents will never collide.
+pub enum IntentScope {
```

Comment:
I think i may be misunderstanding the purpose of "scopes" because having a different "intent" for each type of struct you want to sign seems to defeat the purpose (ie we're just as ridged as we have today). I was expecting something closer to more general intents, ie:
`Consensus`, `Transaction`, `PersonalMessage`, `WalletConnect`.

I'd think there may be some instances where we'd want to sign different types of structs for the same "intent" but maybe i don't know what you're thinking.

Reply:
This is true, btw we should discuss on the Intent naming, I just put something for now (reused the object name) until we agree how we call them.